wrappers/redis/redis.go

```diff
@@ -0,0 +1,169 @@
+	return wrapClient(client, opt)
+}
+
+func wrapClient(client *redis.Client, opt *redis.Options) *redis.Client {
+	host, port := getClientHostPort(opt)
+	client.AddHook(&epsagonHook{
+		host:    host,
+		port:    port,
+		dbIndex: string(opt.DB),
+	})
+	return client
+}
+
+func getClientHostPort(opt *redis.Options) (host, port string) {
+	if opt.Network == "unix" {
+		return "localhost", opt.Addr
+	}
+	host, port, err := net.SplitHostPort(opt.Addr)
+	if err == nil {
+		return host, port
+	}
+	return opt.Addr, ""
+}
+
+func (epsHook *epsagonHook) BeforeProcess(ctx context.Context, cmd redis.Cmder) (context.Context, error) {
+	cmdArgs := safeJsonify(cmd.Args())
+	return epsHook.before(ctx, cmd.Name(), cmdArgs)
+}
+
+func (epsHook *epsagonHook) AfterProcess(ctx context.Context, cmd redis.Cmder) error {
+	var errMsg string
+	if err := cmd.Err(); err != nil {
+		errMsg = err.Error()
+	}
+	return epsHook.after(ctx, cmd.String(), errMsg)
+}
+
+func (epsHook *epsagonHook) BeforeProcessPipeline(ctx context.Context, cmds []redis.Cmder) (context.Context, error) {
+	cmdArgs := safeJsonify(getPiplineCmdArgs(cmds))
+	return epsHook.before(ctx, "Pipeline", cmdArgs)
+}
+
+func (epsHook *epsagonHook) AfterProcessPipeline(ctx context.Context, cmds []redis.Cmder) error {
+	var errMsg string
+	if errors := getPipelineErrors(cmds); len(errors) > 0 {
+		errMsg = safeJsonify(errors)
+
+	}
+	response := safeJsonify(getPipelineResponse(cmds))
+	return epsHook.after(ctx, response, errMsg)
+}
+
+func (epsHook *epsagonHook) before(ctx context.Context, operation, cmdArgs string) (context.Context, error) {
+	metadata := getResourceMetadata(epsHook, cmdArgs)
+	epsHook.event = createEvent(epsHook, operation, metadata)
+	return ctx, nil
+}
+
+func (epsHook *epsagonHook) after(ctx context.Context, response, errMsg string) error {
+	currentTracer := epsagon.ExtractTracer([]context.Context{ctx})
+	if currentTracer == nil {
+		return nil
+	}
+
+	if !currentTracer.GetConfig().MetadataOnly {
+		epsHook.event.Resource.Metadata["redis.response"] = response
+	}
+
+	event := epsHook.event
+	eventEndTime := tracer.GetTimestamp()
+
+	if errMsg != "" {
+		event.ErrorCode = protocol.ErrorCode_EXCEPTION
+		event.Exception = &protocol.Exception{
+			Message:   errMsg,
+			Traceback: string(debug.Stack()),
+			Time:      eventEndTime,
+		}
+	}
+
+	event.Duration = eventEndTime - event.StartTime
+	currentTracer.AddEvent(event)
+	return nil
+}
+
+func createEvent(epsHook *epsagonHook, operation string, metadata map[string]string) *protocol.Event {
+	return &protocol.Event{
+		Id:        "redis-" + uuid.New().String(),
+		StartTime: tracer.GetTimestamp(),
+		Resource: &protocol.Resource{
+			Name:      epsHook.host,
+			Type:      "redis",
+			Operation: operation,
+			Metadata:  metadata,
+		},
+	}
+}
+
+func getResourceMetadata(epsHook *epsagonHook, cmdArgs string) map[string]string {
+	metadata := getConnectionMetadata(epsHook)
+	metadata["Command Arguments"] = cmdArgs
```

Comment:
so this might be saved only if metadataOnly is False (the command arguments, as it might contain a value when there's a set operation)